### PR TITLE
Introduced implicit flow

### PR DESCRIPTION
**Pull request by PeterRydberg**
**Why?**

Generally, the flow to use when the client does not fetch from the UserInfo endpoint directly but rather through a server, is [implicit flow](https://medium.com/@robert.broeckelmann/when-to-use-which-oauth2-grants-and-oidc-flows-ec6a5c00d864). This allows us to to verify the access token (previously, we were using an access code). Implicit flow is also used when not utilizing the client secret.

### CODE REVIEW

**Comment by aslakhol** [comment]
I'm apparently a bit too tired to really get a grip of that blogpost. However the code looks good and everything seems to run. I'll approve it for now, but have at least one other person take a look at it before it's merged. Or at least let me take a look after I've slept if no one else want's to.

**Comment by aslakhol** [comment]
Actually, I'd rather just approve when I've actually considered if we want this change or not. But in any case, I've tested and it seems good.

File: backend/__tests__/controllers/test_application_controller.py [hunk 1]
```diff
@@ -85,7 +85,8 @@ def test_registerNewApplication(mocker, client):
     mimetype = 'application/json'
     headers = {
         'Content-Type': mimetype,
-        'Accept': mimetype
+        'Accept': mimetype,
+        "AccessToken": "Bearer eca448c7-b0f0-487d-98c5-946c3bb29868"
```
Comment (s0lvang) [inline backend/__tests__/controllers/test_application_controller.py]: Use the accessToken from constants.py :)

File: backend/__tests__/testUtils/constants.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 token = "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6ImIwMGE1MDYxYzMxMTQxMjA5Y2RmMjlkYWM3NjYwMTkxMWI2N2M4YWQifQ.eyJpc3MiOiJodHRwczpcL1wvYXV0aC5kYXRhcG9ydGVuLm5vIiwiYXVkIjoiNzdlZTMzY2QtY2M3Zi00YjdhLWJjZTktMjQxYzk2NDU4ZjE0Iiwic3ViIjoiNTVkZTdkNzEtNGEyNS00MTAzLThlNDMtMzVkZjhjMmQ0NzJhIiwiaWF0IjoxNTUxMTgzMTkwLCJleHAiOjE1NTExODY3OTAsImF1dGhfdGltZSI6MTU1MTE4MTYyMX0.L0G7cKUGIoforsRAECsClAKxTqtqCX2QthsQBoZUg1FZAgkDsUivvb1H_RhD-7WrbqAHHvl-hyFD-9qrJth88mjeAoQrpiMGjZNFTKnnLG2NIqXTo9LaHySWLGu30rtswlbua8mbu6R4-3MKGP0sPQriKKU-ngk_6pH93ZNNzBmb8hpDDz5gkzHbhJ1yNhYN2hQ-bvi7tlCBBlyewEw7T-cgKVFN7kBG1p4vf0zcpIiNC9ZgQF8YY6cwTqRUtvkMc88s2GI8E-qiq8YODaD4CPjqvhwHPZotf3J_yFPVLYRuZ8LKjl6QWqvDWqjam_pSFU-XOMDj2gF1-WZhUSUD8A"  # noqa: E501
 accessToken = "eca448c7-b0f0-487d-98c5-946c3bb29868"
+verify_at_hash = True
```
Comment (s0lvang) [inline backend/__tests__/testUtils/constants.py]: haha

File: backend/auth.py [hunk 3]
```diff
@@ -14,35 +14,44 @@
...
 def get_token_auth_header(header):
-    """Obtains the Access Token from the Authorization Header
+    """Obtains the token from a given header
     """
-    auth = request.headers.get(header, None)
-    if not auth:
-        return Response(json.dumps({'error': 'Authorization_header_missing'}), 401)
 
-    parts = auth.split()
+    headerString = ""
```
Comment (s0lvang) [inline backend/auth.py]: It's not nessecary to define the variable before, since it raises an exeption anyways

File: backend/auth.py [hunk 4]
```diff
@@ -14,35 +14,44 @@
...
+    try:
+        headerString = request.headers.get(header, None)
+    except HTTPError:
+        raise HTTPError("Cannot parse header")
 
-    if parts[0].lower() != 'bearer':
-        return Response("{'error': 'Invalid_header'}", 401)
+    if headerString is None:
```
Comment (s0lvang) [inline backend/auth.py]: if not headerString

File: backend/auth.py [hunk 5]
```diff
@@ -14,35 +14,44 @@
...
-    elif len(parts) > 2:
-        return Response("{'error': 'Invalid_header'}", 401)
+    if parts[0].lower() != 'bearer' or len(parts) == 1 or len(parts) > 2:
+        raise TypeError("Header invalid")
 
     token = parts[1]
     return token
 
 
+def eval_access_token():
```
Comment (s0lvang) [inline backend/auth.py]: getAccessToken

File: backend/auth.py [hunk 6]
```diff
@@ -64,7 +73,9 @@ def wrapper(*args, **kwargs):
                         issuer=DOMAIN,
                         audience='77ee33cd-cc7f-4b7a-bce9-241c96458f14',
                         options={"verify_signature": verify,
-                                 "verify_exp": verify}
+                                 "verify_exp": verify,
+                                 "verify_at_hash": verify_at_hash},
```
Comment (s0lvang) [inline backend/auth.py]: We should check this

File: backend/auth.py (hunk 5 above)
Comment (PeterRydberg) [reply backend/auth.py]: Since the function evaluates whether or not to actually use the access token, is it not clearer with this name?

Comment (s0lvang) [reply backend/auth.py]: I don't think it's that clear that it returns the accessToken -_(*-*)_-

File: backend/auth.py (hunk 6 above)
Comment (PeterRydberg) [reply backend/auth.py]: I've created try/except blocks to respond properly if either the ID token or the access token is missing. The ID token is always attempted first, so if both are missing that one will be commented on.

File: backend/auth.py (hunk 5 above)
Comment (PeterRydberg) [reply backend/auth.py]: Ye, did it yo way!

**Pull request merged**